filters/include/pcl/filters/impl/farthest_point_sampling.hpp

```diff
@@ -0,0 +1,107 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef PCL_FILTERS_FARTHEST_POINT_SAMPLING_IMPL_H_
+#define PCL_FILTERS_FARTHEST_POINT_SAMPLING_IMPL_H_
+
+#include <limits>
+#include <algorithm>
+#include <pcl/filters/farthest_point_sampling.h>
+#include <pcl/common/geometry.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (PointCloud &output)
+{
+    int initial_size = input_->points.size();
+    int sample_size = sample_;
+    int max_index;
+    std::vector<int> selected_indices;
+    std::vector<float>distances_to_selected_points;
+    
+    //set random seed
+    std::srand (seed_);
+
+    
+    //check if number of points in cloud is less than requested number of points
+    if (initial_size <= sample_size)
+    {
+        PCL_THROW_EXCEPTION (BadArgumentException,
+                           "Requested number of points is greater than point cloud size!");
+        return;
+    }
+
+    //set initial distances as infinite
+    for (int m = 0; m < initial_size; m++)
+      distances_to_selected_points.push_back(std::numeric_limits<float>::infinity());
+    
+
+    for (int j = 0; j < sample_size; j++)
+    {
+      //pick the first point at random
+      if (j ==0)
+        max_index = rand() % (initial_size -1);
+      else
+      {
+        //select farthest point based on previously calculated distances
+        //since distance is set to -1 for all selected elements,previously selected 
+        //elements are guaranteed to not be selected
+        max_index = std::max_element(distances_to_selected_points.begin(), distances_to_selected_points.end()) - distances_to_selected_points.begin(); 
+      }
+      //set distance to -1 to ignore during max element search
+      distances_to_selected_points[max_index] = -1.0;
+      selected_indices.push_back(max_index);
+      
+      //recompute distances
+      for (int i = 0; i < distances_to_selected_points.size(); i++)
```

Comment:
This is a transformation from `dist[i]` to `i != max ? std::min(dist[i], distance(point[i], point[max])): dist[i]`

But `i` is a simple index here, working on the first `initial_size` elements only. I think something's a little wonky here.

Reply:
So the check for -1.0 here is merely to check if the point is in the set of already selected points. If so, we ignore it. Else (i.e. the point has not yet been selected) we check if the current minimum distance from the selected points to that particular point is greater than the distance from the point that we just selected(`max_index`). If so we update the minimum distance. 

I was going by the algorithm here [Farthest first traversal](https://en.wikipedia.org/wiki/Farthest-first_traversal#Algorithms).

In other words, `i` is iterating over all unselected points, so I'm not sure I see the error here. Please do clarify if I'm missing something.

Reply:
> I'm not sure I see the error here

My confusion was because you used `initial_size` as variable name (which didn't convey that it is the size of the input point cloud). The size is never changed, so why call it `initial_size`?

Anyways, the check for -1 is to confirm that this isn't the same point as the next chosen point. You can rewrite the loop as an algorithm `std::transform` from 0 to `max_index` and `max_index+1` to `initial_size` to convert `dist[i]` to `std::min(dist[i], distance(point[i], point[max]))`. Also [two loops and not one](https://godbolt.org/z/XN3HYf) is at about 30% faster.

Reply:
> My confusion was because you used initial_size as variable name (which didn't convey that it is the size of the input point cloud). The size is never changed, so why call it initial_size?

Yes that's a very good point, when I started writing that I was thinking that it would be changing. I'll rename it to `size`.

> Anyways, the check for -1 is to confirm that this isn't the same point as the next chosen point.

Actually the check for -1 is to see if this point doesn't belong to any of the previously chosen points. I'm not trying to skip over the `max_index` point, I'm trying to skip all of the previously added points, since I'm not removing the points that I've previously selected. So I don't think we can gain that performance boost here. I thought setting the distance as -1.0 would be an easy and efficient way to mark all select points, but it does make it rather look rather confusing, and you can probably suggest a better method.

> Also two loops and not one is about 30% faster.

This is really interesting, I hadn't really thought about this before. I would definitely keep this in mind in the future.

Reply:
Yeah, I got a bit confused in this loop. There indeed isn't any performance boost to be acheived here.